staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -417,28 +392,351 @@ func TestPriorityAndFairnessWithPanicRecoverAndTimeoutFilter(t *testing.T) {
+
+		firstRequestTimesOutPath := "/request/first/time-out-as-designed"
+		secondRequestEnqueuedPath := "/request/second/enqueued-as-designed"
+		firstReqHandlerCompletedCh, firstReqInProgressCh := make(chan struct{}), make(chan struct{})
+		firstReqRoundTripDoneCh, secondReqRoundTripDoneCh := make(chan struct{}), make(chan struct{})
+		requestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			expectMatchingAPFHeaders(t, w, fsName, plName)
+
+			if r.URL.Path == firstRequestTimesOutPath {
+				defer close(firstReqHandlerCompletedCh)
+
+				close(firstReqInProgressCh)
+				<-firstReqRoundTripDoneCh
+
+				// make sure we wait until the caller of the second request returns, this is to
+				// ensure that second request never has a chance to be executed (to avoid flakes)
+				<-secondReqRoundTripDoneCh
+
+				// we expect the timeout handler to have timed out this request by now and any attempt
+				// to write to the response should return a http.ErrHandlerTimeout error.
+				if _, err := w.Write([]byte("foo")); err != http.ErrHandlerTimeout {
+					t.Fatalf("expected error: %#v, but got: %#v", http.ErrHandlerTimeout, err)
+				}
+			}
+
+			if r.URL.Path == secondRequestEnqueuedPath {
+				// we expect the concurrency to be set to 1 and so this request should never be executed.
+				t.Fatalf("expected request to be enqueued: %s", secondRequestEnqueuedPath)
+			}
+		})
+		handler := newHandlerChain(t, requestHandler, controller, userName, requestTimeout)
+
+		server, requestGetter := newHTTP2ServerWithClient(handler)
+		defer server.Close()
+
+		wg := sync.WaitGroup{}
+		wg.Add(2)
+		var firstReqErr, secondReqErr error
+		var resp1, resp2 *http.Response
+		go func() {
+			defer wg.Done()
+			defer close(firstReqRoundTripDoneCh)
+			t.Logf("waiting for the request: %q to time out", firstRequestTimesOutPath)
+			resp1, firstReqErr = requestGetter(firstRequestTimesOutPath)
+		}()
+		go func() {
+			defer wg.Done()
+			defer close(secondReqRoundTripDoneCh)
+
+			// we must wait for the "first" request to start executing first
+			<-firstReqInProgressCh
+			resp2, secondReqErr = requestGetter(secondRequestEnqueuedPath)
+		}()
+
+		wg.Wait()
+		t.Logf("waiting for the inner handler of the request: %q to complete", firstRequestTimesOutPath)
+		<-firstReqHandlerCompletedCh
+
+		// first request is expected to time out.
+		if firstReqErr != nil {
+			t.Fatalf("expected request: %s to succeed, but got error: %#v", firstRequestTimesOutPath, firstReqErr)
+		}
+		if resp1.StatusCode != http.StatusGatewayTimeout {
+			t.Errorf("expected HTTP status code: %d for request: %s, but got: %#v", http.StatusGatewayTimeout, firstRequestTimesOutPath, resp1)
+		}
+
+		// second request is expected to be rejected
+		if secondReqErr != nil {
+			t.Fatalf("expected request: %s to succeed, but got error: %#v", secondRequestEnqueuedPath, secondReqErr)
+		}
+		if resp2.StatusCode != http.StatusGatewayTimeout {
```

Comment:
This is currently checking specifically for implementation behavior that is admittedly only an approximation of the ideal.  As far as I am concerned it would be fine for this _test_ to both accept the current approximate behavior and accept the ideal behavior.